vm/src/stdlib/zlib.rs

```diff
@@ -477,34 +480,36 @@ mod decl {
     const CHUNKSIZE: usize = u32::MAX as usize;
 
     impl CompressInner {
-        fn save_unconsumed_input(&mut self, data: &[u8], orig_in: u64) {
-            let leftover = &data[(self.compress.total_in() - orig_in) as usize..];
+        fn save_unconsumed_input(&mut self, data: &[u8], cur_in: usize) {
+            let leftover = &data[cur_in..];
             self.unconsumed.extend_from_slice(leftover);
         }
 
         fn compress(&mut self, data: &[u8], vm: &VirtualMachine) -> PyResult<Vec<u8>> {
-            let orig_in = self.compress.total_in();
+            let orig_in = self.compress.total_in() as usize;
+            let mut cur_in = 0;
             let unconsumed = std::mem::take(&mut self.unconsumed);
             let mut buf = Vec::new();
 
             'outer: for chunk in unconsumed.chunks(CHUNKSIZE).chain(data.chunks(CHUNKSIZE)) {
-                loop {
+                while cur_in < chunk.len() {
                     buf.reserve(DEF_BUF_SIZE);
                     let status = self
                         .compress
-                        .compress_vec(chunk, &mut buf, FlushCompress::None)
+                        .compress_vec(&chunk[cur_in..], &mut buf, FlushCompress::None)
                         .map_err(|_| {
-                            self.save_unconsumed_input(data, orig_in);
+                            self.save_unconsumed_input(data, cur_in);
```

Comment:
e.g. here this can just be one line as well, `self.unconsumed.extend_from_slice(&data[cur_in..]);`